### PR TITLE
Feature/setup+auth service

### PR DESCRIPTION
**Pull request by Limopit**

### CODE REVIEW

File: Backend/CourseAppUserService/CourseAppUserService_Application/UserCreatedCourse/Commands/CreateUserCreatedCourse/CreateUserCreatedCourseCommand.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using MediatR;
+
+namespace CourseAppUserService_Application.UserCreatedCourse.Commands.CreateUserCreatedCourse;
+
+public class CreateUserCreatedCourseCommand: IRequest<Guid>
+{
+    public required string Email { get; set; }
+    public required string CourseId { get; set; }
+    public required string ApprovementStatus { get; set; } = "Pending";
```
Comment (Pavelabramovich) [inline Backend/CourseAppUserService/CourseAppUserService_Application/UserCreatedCourse/Commands/CreateUserCreatedCourse/CreateUserCreatedCourseCommand.cs]: Mb replace status type with enum? Or will be many possible statuses?

File: Backend/CourseAppUserService/CourseAppUserService_WebAPI/Middleware/CustomExceptionHandler.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                break;
+        }
+
+        context.Response.ContentType = "application/json";
+        context.Response.StatusCode = (int)code;
+
+        if (result == string.Empty)
+        {
+            result = JsonSerializer.Serialize(new { err = e.Message });
+        }
```
Comment (ArtyomTyshkevich) [inline Backend/CourseAppUserService/CourseAppUserService_WebAPI/Middleware/CustomExceptionHandler.cs:54]: Avoid using abbreviations like e or err

File: Backend/CourseAppUserService/CourseAppUserService_Domain/CourseAppUserService_Domain.csproj
Comment (ArtyomTyshkevich) [inline Backend/CourseAppUserService/CourseAppUserService_Domain/CourseAppUserService_Domain.csproj:1]: Consider organizing your Entities classes into a separate folder

File: Backend/CourseAppNotificationService/CourseAppNotificationService/obj/CourseAppNotificationService.csproj.nuget.dgspec.json
Comment (correntis) [inline Backend/CourseAppNotificationService/CourseAppNotificationService/obj/CourseAppNotificationService.csproj.nuget.dgspec.json:1]: Add `/bin` `/obj` `.http` and mb `.sln` to **.gitignore**

File: Backend/CourseAppUserService/CourseAppUserService_Application/Common/Behavior/ValidationBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+using MediatR;
+
+namespace CourseAppUserService_Application.Common.Behavior;
+
+public class ValidationBehavior<TRequest, TResponse>(IEnumerable<IValidator<TRequest>> validators)
+    : IPipelineBehavior<TRequest, TResponse>
+    where TRequest : notnull
+{
+    public Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse> next,
+        CancellationToken cancellationToken)
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Application/Common/Behavior/ValidationBehavior.cs]: Each parameter on a new line, or all in one line

File: Backend/CourseAppUserService/CourseAppUserService_Application/Common/Exceptions/UserValidationException.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+using FluentValidation;
+using FluentValidation.Results;
+
+namespace CourseAppUserService_Application.Common.Exceptions;
+
+public class UserValidationException: ValidationException
+{
+    public List<string> errors { get; set; }
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Application/Common/Exceptions/UserValidationException.cs]: Properties names in CamelCase
```suggestion
    public List<string> Errors { get; set; }
```

File: Backend/CourseAppUserService/CourseAppUserService_Application/UserCreatedCourse/Queries/GetUserCreatedCourses/UserCreatedCourseVm.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+namespace CourseAppUserService_Application.UserCreatedCourse.Queries.GetUserCreatedCourses;
+
+public class UserCreatedCourseVm
+{
+    public IList<UserCreatedCourseDto> UserCreatedCourses { get; set; }
+
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Application/UserCreatedCourse/Queries/GetUserCreatedCourses/UserCreatedCourseVm.cs]: Extra line

File: Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/LoginUser/LoginUserCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        if (user == null)
+        {
+            throw new Exception("Wrong email or password");
+        }
+        
+        var isPasswordValid = await unitOfWork.Users.CheckPassword(user, request.Password);
+        if (!isPasswordValid)
+        {
+            throw new Exception("Wrong email or password");
+        }
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/LoginUser/LoginUserCommandHandler.cs]: Conside add exceptions like `InvalidPasswordEx `and `InvalidEmailEx`. You will be able to return more suitable status codes and in frontend show in which specific field there was an error

File: Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/UpdateUserPassword/UpdateUserPasswordCommand.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+using MediatR;
+
+namespace CourseAppUserService_Application.Users.Commands.UpdateUserPassword;
+
+public class UpdateUserPasswordCommand: IRequest
+{
+    public string PrevPassword { get; set; }
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/UpdateUserPassword/UpdateUserPasswordCommand.cs]: Maybe `OldPassword` would be better? And you won't have to use the abbreviation

File: Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/BaseRepository.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        await _dbSet.AddAsync(entity, token);
+    }
+    
+    public Task UpdateAsync(T entity)
+    {
+        _dbSet.Update(entity);
+        return Task.CompletedTask;
+    }
+
+    public Task RemoveEntity(T entity)
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/BaseRepository.cs]: RemoveEntity`Async`

File: Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/RefreshTokenRepository.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
+using CourseAppUserService_Application.Interfaces.Repositories;
+using CourseAppUserService_Domain;
+using Microsoft.EntityFrameworkCore;
+
+namespace CourseAppUserService_Persistance.DbPatterns.Repositories;
+
+public class RefreshTokenRepository(UserServiceDbContext libraryDbContext) : IRefreshTokenRepository
+{
+    public async Task<RefreshToken?> ValidateRefreshToken(string refreshToken)
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/RefreshTokenRepository.cs]: ValidateRefreshToken`Async`

File: Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/RefreshTokenRepository.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        return await libraryDbContext.RefreshTokens
+            .Where(r => r.Token == refreshToken && r.Expires > DateTime.UtcNow)
+            .FirstOrDefaultAsync();
+    }
+
+    public async Task<RefreshToken?> RevokeToken(string refreshToken)
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/RefreshTokenRepository.cs]: RevokeToken`Async`

File: Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/UserRepository.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    UserManager<User> userManager,
+    RoleManager<IdentityRole> roleManager)
+    : BaseRepository<User>(context), IUserRepository
+{
+    public async Task<IdentityResult?> AddUserAsync(User user, string password)
+    {
+        return await userManager.CreateAsync(user, password);
+    }
+
+    public async Task<User?> FindUserByEmail(string email)
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DbPatterns/Repositories/UserRepository.cs]: FindUserByEmail`Async`

File: Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DependencyInjection.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        services.AddScoped<IUserCreatedCourseRepository, UserCreatedCourseRepository>();
+        services.AddScoped<IRefreshTokenRepository, RefreshTokenRepository>();
+        services.AddScoped<IUnitOfWork, UnitOfWork>();
+        services.AddScoped<ITokenService, TokenService>();
+        services.AddScoped<IHttpContextService, HttpContextService>();
+        services.AddTransient<IMapperService, MapperService>();
+    
+        return services;
+    }
+
```
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_Infrastructure/DependencyInjection.cs]: Empty line

File: Backend/CourseAppUserService/CourseAppUserService_WebAPI/Controllers/AuthController.cs
Comment (correntis) [inline Backend/CourseAppUserService/CourseAppUserService_WebAPI/Controllers/AuthController.cs:1]: Unnecessary duplication of the http method in the endpoint name. 
Following REST:
```
GET /users
DELETE /users
etc.
```

**Review by Pavelabramovich — APPROVED** [review]

**Review by correntis — APPROVED** [review]
Conside split `UsersController` to `AuthController` and `UsersController`

**Review by ArtyomTyshkevich — APPROVED** [review]

**Review by correntis — APPROVED** [review]

File: Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/RegisterUser/RegisterUserCommandHandler.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    public async Task<string> Handle(RegisterUserCommand request, CancellationToken cancellationToken)
+    {
+        var user = new User { 
+            UserName = request.Email,
+            Email = request.Email,
+            FirstName = request.FirstName, 
+            LastName = request.LastName
+        };
+        
+        if (!await unitOfWork.Users.UserRoleExistsAsync(request.Role))
```
Comment (polashkind) [inline Backend/CourseAppUserService/CourseAppUserService_Application/Users/Commands/RegisterUser/RegisterUserCommandHandler.cs]: consider separate `!await unitOfWork.Users.UserRoleExistsAsync(request.Role)` to variable

**Review by polashkind — APPROVED** [review]

**Pull request merged**